File: use/cross/protobuf/cpp/example/hello/hello.cpp
#include <stdio.h>
#include <iostream>
#include <map>
#include <functional>
#include <google/protobuf/message.h>
#include "gen/proto/gameobject.pb.h"
#include "gen/proto/food.pb.h"
#include "codec.h"
#include "msg_handler.h"

using namespace google::protobuf;
using namespace gen::proto;

class Receiver
{
public:
	void OnMessage(std::shared_ptr<Message> &msg)
	{
		std::cout << std::endl << "========" << std::endl;
		std::cout << "on message: " << msg->GetTypeName() << std::endl;
	}

	void OnMessagePosition(std::shared_ptr<Position> &msg)
	{
		std::cout << std::endl << "========" << std::endl;
		std::cout << "OnMessagePosition: " << std::endl
			<< msg->x() << ", "
			<< msg->y() << ", "
			<< msg->z() << std::endl;
	}
	void OnMessageRotation(std::shared_ptr<Rotation> &msg)
	{
		std::cout << std::endl << "========" << std::endl;
		std::cout << "OnMessageRotation: " << std::endl
			<< msg->x_axis() << ", "
			<< msg->y_axis() << ", "
			<< msg->z_axis() << std::endl;
	}
	void OnMessageTransform(std::shared_ptr<Transform> &msg)
	{
		std::cout << std::endl << "========" << std::endl;
		std::cout << "OnMessageTransform: " << std::endl
			<< msg->position().x() << ", "
			<< msg->position().y() << ", "
			<< msg->position().z() << std::endl
			<< msg->rotation().x_axis() << ", "
			<< msg->rotation().y_axis() << ", "
			<< msg->rotation().z_axis() << std::endl
			<< msg->scale().x_scale() << ", "
			<< msg->scale().y_scale() << ", "
			<< msg->scale().z_scale() << std::endl;
	}
	void OnMessageLunch(std::shared_ptr<Lunch> &msg)
	{
		std::cout << std::endl << "========" << std::endl;
		std::cout << "OnMessageLunch: " << std::endl;
		for (int i = 0; i < msg->foods_size(); ++i)
		{
			switch (msg->foods(i).food_oneof_case())
			{
			case Food::FoodOneofCase::kRice:
			{
				auto rice = msg->foods(i).rice();
				std::cout << i << ") rice: " << rice.origin_place() << std::endl;
			}break;
			case Food::FoodOneofCase::kDumpling:
			{
				auto dumpling = msg->foods(i).dumpling();
				std::cout << i << ") dumpling: " << dumpling.stuffing() << ", " << dumpling.taste() << std::endl;
			}break;
			case Food::FoodOneofCase::kNoodle:
			{
				auto noodle = msg->foods(i).noodle();
				std::cout << i << ") noodle: " << noodle.material() << ", " << noodle.taste() << std::endl;
			}break;
			case Food::FoodOneofCase::FOOD_ONEOF_NOT_SET:
			{
				std::cout << i << ") not set: " << std::endl;
			}break;
			}
		}
	}

	void Register()
	{
		handler_.RegisterDefault<Message>(std::bind(&Receiver::OnMessage, this, std::placeholders::_1));
		
		handler_.Register<Position>(std::bind(&Receiver::OnMessagePosition, this, std::placeholders::_1));
		handler_.Register<Rotation>(std::bind(&Receiver::OnMessageRotation, this, std::placeholders::_1));
		handler_.Register<Transform>(std::bind(&Receiver::OnMessageTransform, this, std::placeholders::_1));
		handler_.Register<Lunch>(std::bind(&Receiver::OnMessageLunch, this, std::placeholders::_1));
	}

	int32_t Parse(char *bytes)
	{
		int32_t total_len = 0;
		Message *msg = codec_.Deserialize(bytes, total_len);
		if (msg)
		{
			auto sp = std::shared_ptr<Message>(msg);
			handler_.OnMessage(sp);
		}

		return total_len;
	}

private:
	Codec codec_;
	ProtoMessageHandler handler_;
};

void baseSample()
{
	Receiver receiver;
	Codec codec;
	char *bytes = nullptr;
	int32_t total_len = 0;

	receiver.Register();

	// position
	Position pos;
	pos.set_x(5.0);
	pos.set_y(1.0);
	pos.set_z(6.0);
	bytes = codec.Serialize(&pos, total_len);
	receiver.Parse(bytes);
	free(bytes);

	// rotation
	Rotation rot;
	rot.set_x_axis(3.14f);
	rot.set_y_axis(0.0f);
	rot.set_z_axis(1.57f);
	bytes = codec.Serialize(&rot, total_len);
	receiver.Parse(bytes);
	free(bytes);

	// scale
	Scale scale;
	scale.set_x_scale(1.0f);
	scale.set_y_scale(1.0f);
	scale.set_z_scale(1.0f);
	bytes = codec.Serialize(&scale, total_len);
	receiver.Parse(bytes);
	free(bytes);

	// transform
	Transform transform;
	*transform.mutable_position() = pos;
	*transform.mutable_rotation() = rot;
	bytes = codec.Serialize(&transform, total_len);
	receiver.Parse(bytes);
	free(bytes);

	// lunch
	Lunch lunch;

	lunch.add_foods()->mutable_rice()->set_origin_place("fujian");

	auto noddle = lunch.add_foods()->mutable_noodle();
	noddle->set_material("flour");
	noddle->set_taste(Taste::NO_SPICY);

	auto dumpling = lunch.add_foods()->mutable_dumpling();
	dumpling->set_stuffing("Leek pork");
	dumpling->set_taste(Taste::SPICY);

	lunch.add_foods();

	bytes = codec.Serialize(&lunch, total_len);
	receiver.Parse(bytes);
	free(bytes);
}

void writeSample()
{
	Codec codec;
	char *bytes = nullptr;
	int32_t total_len = 0;

	FILE *fp = fopen("msg_bytes", "wb");
	if (fp == nullptr)
	{
		return;
	}

	// position
	Position pos;
	pos.set_x(5.0);
	pos.set_y(1.0);
	pos.set_z(6.0);
	bytes = codec.Serialize(&pos, total_len);
	fwrite(bytes, 1, total_len, fp);
	free(bytes);

	// rotation
	Rotation rot;
	rot.set_x_axis(3.14f);
	rot.set_y_axis(0.0f);
	rot.set_z_axis(1.57f);
	bytes = codec.Serialize(&rot, total_len);
	fwrite(bytes, 1, total_len, fp);
	free(bytes);

	// scale
	Scale scale;
	scale.set_x_scale(1.0f);
	scale.set_y_scale(1.0f);
	scale.set_z_scale(1.0f);
	bytes = codec.Serialize(&scale, total_len);
	fwrite(bytes, 1, total_len, fp);
	free(bytes);

	// transform
	Transform transform;
	*transform.mutable_position() = pos;
	*transform.mutable_rotation() = rot;
	bytes = codec.Serialize(&transform, total_len);
	fwrite(bytes, 1, total_len, fp);
	free(bytes);

	// lunch
	Lunch lunch;

	lunch.add_foods()->mutable_rice()->set_origin_place("fujian");

	auto noddle = lunch.add_foods()->mutable_noodle();
	noddle->set_material("flour");
	noddle->set_taste(Taste::NO_SPICY);

	auto dumpling = lunch.add_foods()->mutable_dumpling();
	dumpling->set_stuffing("Leek pork");
	dumpling->set_taste(Taste::SPICY);

	lunch.add_foods();

	bytes = codec.Serialize(&lunch, total_len);
	fwrite(bytes, 1, total_len, fp);
	free(bytes);

	fclose(fp);
}

void readSample()
{
	Receiver receiver;
	receiver.Register();

	FILE *fp = fopen("msg_bytes", "rb");
	if (fp == nullptr)
	{
		return;
	}

	fseek(fp, 0, SEEK_END);
	long file_len = ftell(fp);
	fseek(fp, 0, SEEK_SET);

	void *bytes = malloc(file_len);
	fread(bytes, file_len, 1, fp);

	int32_t accum_len = 0;
	while (true)
	{
		char *p = (char*)bytes + accum_len;
		int32_t parse_len = receiver.Parse(p);
		accum_len += parse_len;
		if (accum_len >= file_len)
		{
			free(bytes);
			break;
		}
	}
}

int main(int argc, char *argv[])
{
	if (argc == 1)
	{
		baseSample();
	}
	else if (argc >= 2)
	{
		if (argv[1][0] == 'r')
		{
			readSample();
		}
		else if (argv[1][0] == 'w')
		{
			writeSample();
		}
		else
		{
			std::cout << "enter: " << argv[0] << " [r|w]" << std::endl;
		}
	}

	return 0;
}

File: use/cross/protobuf/java/src/main/java/com/muggle/helloproto/Hello.java
package com.muggle.helloproto;

import com.google.protobuf.Message;
import com.muggle.proto.Codec;
import gen.proto.FoodOuterClass;
import gen.proto.Gameobject;
import javafx.geometry.Pos;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class Hello {
    static public void main(String[] args) {
        if (args.length != 0) {
            if (args[0].equals("r")) {
                readSample();
            } else if (args[0].equals("w")) {
                writeSample();
            } else {
                System.out.println("enter: command [r|w]");
            }
        } else {
            baseSample();
        }
    }

    static private void baseSample() {
        MyReceiver receiver = new MyReceiver();
        Codec codec = new Codec();
        byte[] bytes = null;

        receiver.Register();

        // position
        Gameobject.Position position = Gameobject.Position.newBuilder()
                .setX(5.0f).setY(1.0f).setZ(6.0f).build();
        bytes = codec.Serialize(position);
        receiver.Parse(bytes);

        // rotation
        Gameobject.Rotation rotation = Gameobject.Rotation.newBuilder()
                .setXAxis(3.14f).setYAxis(0.0f).setZAxis(1.57f).build();
        bytes = codec.Serialize(rotation);
        receiver.Parse(bytes);

        // Transform
        Gameobject.Transform transform = Gameobject.Transform.newBuilder()
                .setPosition(position)
                .setRotation(rotation)
                .build();
        bytes = codec.Serialize(transform);
        receiver.Parse(bytes);

        // lunch
        FoodOuterClass.Lunch.Builder lunchBuilder = FoodOuterClass.Lunch.newBuilder();

        FoodOuterClass.Rice rice = FoodOuterClass.Rice.newBuilder().setOriginPlace("fujian").build();
        FoodOuterClass.Food foodRice = FoodOuterClass.Food.newBuilder().setRice(rice).build();
        lunchBuilder.addFoods(foodRice);

        FoodOuterClass.Noodle noodle = FoodOuterClass.Noodle.newBuilder()
                .setMaterial("flour")
                .setTaste(FoodOuterClass.Taste.NO_SPICY)
                .build();
        FoodOuterClass.Food foodNoodle = FoodOuterClass.Food.newBuilder().setNoodle(noodle).build();
        lunchBuilder.addFoods(foodNoodle);

        FoodOuterClass.Dumpling dumpling = FoodOuterClass.Dumpling.newBuilder()
                .setStuffing("Leek pork")
                .setTaste(FoodOuterClass.Taste.SPICY)
                .build();
        FoodOuterClass.Food foodDumpling = FoodOuterClass.Food.newBuilder().setDumpling(dumpling).build();
        lunchBuilder.addFoods(foodDumpling);

        lunchBuilder.addFoods(FoodOuterClass.Food.newBuilder().build());
        bytes = codec.Serialize(lunchBuilder.build());
        receiver.Parse(bytes);

    }

    static private void readSample() {
        File f = new File("msg_bytes");
        long len = f.length();

        try (BufferedInputStream reader = new BufferedInputStream(new FileInputStream(f))){
            byte[] bytes = new byte[(int)len];
            reader.read(bytes);

            MyReceiver receiver = new MyReceiver();
            Codec codec = new Codec();
            receiver.Register();

            ByteBuffer byteBuffer = ByteBuffer.allocate(4);
            byteBuffer.order(ByteOrder.BIG_ENDIAN);

            int accum = 0;
            while (true) {
                byteBuffer.clear();
                byteBuffer.put(bytes, accum, 4);
                byteBuffer.clear();
                int messageLen = byteBuffer.getInt();

                byte[] messageBytes = new byte[messageLen];
                for (int i = 0; i < messageLen; ++i) {
                    messageBytes[i] = bytes[accum + i];
                }
                receiver.Parse(messageBytes);
                accum += messageLen;
                if (accum >= len) {
                    break;
                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    static private void writeSample() {
        Codec codec = new Codec();
        byte[] bytes = null;
        try (BufferedOutputStream writer = new BufferedOutputStream(new FileOutputStream("msg_bytes"))) {
            // position
            Gameobject.Position position = Gameobject.Position.newBuilder()
                    .setX(5.0f).setY(1.0f).setZ(6.0f).build();
            bytes = codec.Serialize(position);
            writer.write(bytes);

            // rotation
            Gameobject.Rotation rotation = Gameobject.Rotation.newBuilder()
                    .setXAxis(3.14f).setYAxis(0.0f).setZAxis(1.57f).build();
            bytes = codec.Serialize(rotation);
            writer.write(bytes);

            // scale
            Gameobject.Scale scale = Gameobject.Scale.newBuilder()
                    .setXScale(1.0f).setYScale(1.0f).setZScale(1.0f).build();
            bytes = codec.Serialize(scale);
            writer.write(bytes);

            // Transform
            Gameobject.Transform transform = Gameobject.Transform.newBuilder()
                    .setPosition(position)
                    .setRotation(rotation)
                    .build();
            bytes = codec.Serialize(transform);
            writer.write(bytes);

            // lunch
            FoodOuterClass.Lunch.Builder lunchBuilder = FoodOuterClass.Lunch.newBuilder();

            FoodOuterClass.Rice rice = FoodOuterClass.Rice.newBuilder().setOriginPlace("fujian").build();
            FoodOuterClass.Food foodRice = FoodOuterClass.Food.newBuilder().setRice(rice).build();
            lunchBuilder.addFoods(foodRice);

            FoodOuterClass.Noodle noodle = FoodOuterClass.Noodle.newBuilder()

                    .setMaterial("flour")
                    .setTaste(FoodOuterClass.Taste.NO_SPICY)
                    .build();
            FoodOuterClass.Food foodNoodle = FoodOuterClass.Food.newBuilder().setNoodle(noodle).build();
            lunchBuilder.addFoods(foodNoodle);

            FoodOuterClass.Dumpling dumpling = FoodOuterClass.Dumpling.newBuilder()
                    .setStuffing("Leek pork")
                    .setTaste(FoodOuterClass.Taste.SPICY)
                    .build();
            FoodOuterClass.Food foodDumpling = FoodOuterClass.Food.newBuilder().setDumpling(dumpling).build();
            lunchBuilder.addFoods(foodDumpling);

            lunchBuilder.addFoods(FoodOuterClass.Food.newBuilder().build());
            bytes = codec.Serialize(lunchBuilder.build());
            writer.write(bytes);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }


    }
}
